readium/navigator/src/main/java/org/readium/r2/navigator/media2/MediaSessionNavigator.kt

```diff
@@ -0,0 +1,327 @@
+
+
+    data class Configuration(
+        val positionRefreshRate: Double = 2.0,  // Hz
+        val skipForwardInterval: Duration = 30.seconds,
+        val skipBackwardInterval: Duration = 30.seconds,
+    )
+
+    @ExperimentalTime
+    data class Playback(
+        val state: State,
+        val rate: Double,
+        val currentIndex: Int,
+        val currentLink: Link,
+        val currentPosition: Duration,
+        val bufferedPosition: Duration,
+    ) {
+
+        enum class State {
+            Playing,
+            Paused,
+            Finished,
+            Error
+        }
+    }
+
+    sealed class Exception : kotlin.Exception()
+
+    /*
+     * Compatibility
+     */
+
+    private fun launchAndRun(runnable: suspend () -> Unit, callback: () -> Unit) =
+        coroutineScope.launch { runnable() }.invokeOnCompletion { callback() }
+
+    override fun go(locator: Locator, animated: Boolean, completion: () -> Unit): Boolean {
+        launchAndRun({ go(locator) }, completion)
+        return true
+    }
+
+    override fun go(link: Link, animated: Boolean, completion: () -> Unit): Boolean {
+        launchAndRun({ go(link) }, completion)
+        return true
+    }
+
+    override fun goForward(animated: Boolean, completion: () -> Unit): Boolean {
+        launchAndRun({ goForward() }, completion)
+        return true
+    }
+
+    override fun goBackward(animated: Boolean, completion: () -> Unit): Boolean {
+        launchAndRun({ goBackward() }, completion)
+        return true
+    }
+
+    companion object {
+
+        suspend fun create(
+            context: Context,
+            publication: Publication,
+            initialLocator: Locator?,
+            configuration: Configuration = Configuration(),
+            player: SessionPlayer = ExoPlayerFactory().createPlayer(context, publication)
+        ): Try<MediaSessionNavigator, Exception> {
+
+            val positionRefreshDelay = (1.0 / configuration.positionRefreshRate).seconds
+            val callback = SessionPlayerCallback(positionRefreshDelay)
+            val callbackExecutor = Executors.newSingleThreadExecutor()
+            player.registerPlayerCallback(callbackExecutor, callback)
+
+            val facade = SessionPlayerFacade(player)
+            return preparePlayer(publication, facade)
+                // Ignoring failure to set initial locator
+                .onSuccess { goInitialLocator(publication, initialLocator, facade) }
+                // Player must be ready to play when MediaNavigator's constructor is called.
+                .map { MediaSessionNavigator(publication, facade, callback, configuration) }
+        }
+
+        private suspend fun preparePlayer(
+            publication: Publication, player:
+            SessionPlayerFacade
+        ): Try<Unit, Exception> {
+            val playlist = publication.readingOrder.toPlayList()
+            val metadata = publicationMetadata(publication)
```

Comment:
I don't know. Are media style notifications limited to player's metadata? If not, apps are free to use publication's metadata when they build up their notifications.

Reply:
This looks to be the default one, as mentioned [here](https://developer.android.com/guide/topics/media/media-controls#mediasession_implementation).  I'm wondering whether it has a default at all for Android 10 and below though.

But otherwise it would be the developer's responsibility to override `onUpdateNotification` when extending `MediaSessionService` in their app to display their own notification control.  However, the metadata in that notification needs to keep in sync with what is in the MediaSession in the Navigator, and we should ensure that is possible.  We should also ensure developer's can control aspects of playback so they can add their own actions in the notification controls.

Since the metadata we inject in the Navigator is also accessible from `MediaSession` being passed into `onUpdateNotification`, I would suggest we at least add the following:

- METADATA_KEY_ARTIST (a default):  Author(s) of the publication
- METADATA_KEY_TITLE (a default):  Title of the publication
- METADATA_KEY_ALBUM_ART:  Bitmap of the publication cover
- METADATA_KEY_DISPLAY_SUBTITLE:  The title of the current chapter

And consider adding:

- METADATA_KEY_NUM_TRACKS:  Number of chapters of the publication
- METADATA_KEY_AUTHOR:  Author(s) of the publication

Reply:
Done.